horizomer/utils/tree.py

```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env python
+
+# ----------------------------------------------------------------------------
+# Copyright (c) 2015--, The Horizomer Development Team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+#
+# functions relevant to phylogenetic tree operations
+#
+
+from skbio import TreeNode
+
+
+def compare_topology(node1, node2):
```

Comment:
This function here poses an interesting algorithmic problem, therefore I had some additional thoughts about it. Not sure if it is worth to explore them in thus PR, maybe better ensure that the current code is doing what is shall do and later trigger a new PR?

At some point you depend on sorting the children nodes alphanumerically with regards to their names. Can it happen, that two children have identical names? If yes, if you recurse further down, how can you ensure that you are looking at the right child? I.e. is your sorting "stable"?

I agree with @mortonjt that recursions are not the strength of python and should be avoided as much as possible. However, due to the re-ordering of child nodes, I don't see a direct way to use pre- or postorder traversals for this use case.

However, if we assume every node has a unique name, we might solve the problem with two traversals per tree:
1) preorder traversal: reorder children by lexicographic sorting
2) preorder traversal: concatenate all node names (delimited with a char not in node names like '#') into one long string
3) compare those strings for tree1 and tree2.

Reply:
can't stop thinking about this problem. Correct me if I am wrong, but for topology comparison you are basically only interested in the parent-child relationships, right? If this assumption holds you could do the following:
1) preorder traverse tree1 and store all (parent.name: child.name) relations in one dict
2) preorder traverse tree2 and for every found parent.name: child.name relation, check if it is contained in the dict and keys and values match.
Thus, you avoid explicit sorting and recursion

Reply:
Hi @sjanssen2 thanks for your very insightful thoughts!

Yes it can happen that two children have identical names. Actually for many situations, internal nodes don't have labels (or labels == bootstrap values). In those circumstances, this function no longer works, but scikit-bio's compare_subsets or compare_rfd are better choices. Therefore, the current function assumes that all internal nodes have unique node IDs.

Your idea of concatenated string comparison is interesting! At my first glance it should be much more efficient. Regarding string comparison I think we may design very carefully. For example, to my understanding, for tree `((d,e)b,(f,g)c)a;`, the concatenated string will be `abdecfg`. However, for tree `((d,e,c)b,(g)f)a;` which has a different topology, the string will be the same. Maybe I didn't understand your idea correctly?

Reply:
@sjanssen2 I just tested your last idea by this:
```
    dic1 = {node.name: set([x.name for x in node.children])
            for node in node1.non_tips()}
    dic2 = {node.name: set([x.name for x in node.children])
            for node in node2.non_tips()}
    return dic1 == dic2
```
But it didn't generate the correct result. Maybe I didn't do it right...

Reply:
@sjanssen2 okay I think I figured out two solutions, one testing node-parent and the other testing node-children. Both passed unit tests:
1
```
dic1, dic2 = ({node.name: set([child.name for child in node.children])
              for node in tree.traverse() if not node.is_tip()}
              for tree in (tree1, tree2))
    return dic1 == dic2
```
2
```
    dic1, dic2 = ({node.name: node.parent.name
                  for node in tree.traverse() if not node.is_root()}
                  for tree in (tree1, tree2))
    return dic1 == dic2
```